src/dbnode/storage/shard.go

```diff
@@ -2156,18 +2169,36 @@ func (s *dbShard) removeAnyFlushStatesTooEarly(tickStart time.Time) {
 
 func (s *dbShard) CleanupExpiredFileSets(earliestToRetain time.Time) error {
 	filePathPrefix := s.opts.CommitLogOptions().FilesystemOptions().FilePathPrefix()
-	multiErr := xerrors.NewMultiError()
-	expired, err := s.filesetBeforeFn(filePathPrefix, s.namespace.ID(), s.ID(), earliestToRetain)
+	expired, err := s.filesetPathsBeforeFn(filePathPrefix, s.namespace.ID(), s.ID(), earliestToRetain)
 	if err != nil {
-		detailedErr :=
-			fmt.Errorf("encountered errors when getting fileset files for prefix %s namespace %s shard %d: %v",
-				filePathPrefix, s.namespace.ID(), s.ID(), err)
-		multiErr = multiErr.Add(detailedErr)
+		return fmt.Errorf("encountered errors when getting fileset files for prefix %s namespace %s shard %d: %v",
+			filePathPrefix, s.namespace.ID(), s.ID(), err)
 	}
-	if err := s.deleteFilesFn(expired); err != nil {
-		multiErr = multiErr.Add(err)
+
+	return s.deleteFilesFn(expired)
+}
+
+func (s *dbShard) CleanupCompactedFileSets() error {
+	filePathPrefix := s.opts.CommitLogOptions().FilesystemOptions().FilePathPrefix()
+	filesets, err := s.filesetsFn(filePathPrefix, s.namespace.ID(), s.ID())
+	if err != nil {
+		return fmt.Errorf("encountered errors when getting fileset files for prefix %s namespace %s shard %d: %v",
+			filePathPrefix, s.namespace.ID(), s.ID(), err)
 	}
-	return multiErr.FinalError()
+	// Get a snapshot of all states here to prevent constantly getting/releasing
+	// locks in a tight loop below.
+	blockStates := s.BlockStatesSnapshot()
+	// Filter without allocating by using same backing array.
+	compacted := filesets[:0]
```

Comment:
I'd probably call this `toDelete` or something

Reply:
also this is sketching me out lol can we just allocate a new slice. This is only once per shard so its not so bad

Reply:
Alllllright.